# Sprocket Central Pty Ltd Data Set

#### importing necessery libraries for future analysis of the dataset

In [1]:
import numpy as np
import statistics as st
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from pandas import ExcelWriter
from pandas import ExcelFile

#### Creating data frame

In [2]:
df = pd.read_excel("KPMG.xlsx")

#### creating separate variable for data set

In [3]:
transaction = pd.read_excel("KPMG.xlsx",sheet_name = 1,skiprows = 1)
CustomerDemographic = pd.read_excel("KPMG.xlsx", sheet_name = 3 , skiprows = 1)
CustomerAddress = pd.read_excel("KPMG.xlsx", sheet_name = 4, skiprows = 1)

#### -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

#### checking transaction data set

In [4]:
transaction.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [5]:
transaction.shape

(20000, 13)

In [6]:
transaction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [7]:
transaction.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

In [8]:
transaction.brand.value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

#### checking null values in the transaction data frame

In [9]:
transaction.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

#### Checking for Brand Column

In [10]:
transaction.brand.value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

#### Dropping brand null value


In [11]:
transaction_1 = transaction[transaction['brand'].isnull()]

In [12]:
transaction.drop(transaction_1.index,inplace = True)

#### Finding mode for order_online 


In [13]:
mode_online_data =  transaction.online_order.mode()[0]
mode_online_data

1.0

#### Replacing the null value with mode_online_data

In [14]:
transaction['online_order'] = transaction['online_order'].fillna(mode_online_data)

#### Checking null value 


In [15]:
transaction.isnull().sum()

transaction_id             0
product_id                 0
customer_id                0
transaction_date           0
online_order               0
order_status               0
brand                      0
product_line               0
product_class              0
product_size               0
list_price                 0
standard_cost              0
product_first_sold_date    0
dtype: int64

#### ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

### lets check CustomerDemographic data frame

In [16]:
CustomerDemographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [17]:
CustomerDemographic.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure'],
      dtype='object')

In [18]:
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [19]:
CustomerDemographic.shape

(4000, 13)

In [20]:
CustomerDemographic.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

#### Lets Check unique value

In [21]:
CustomerDemographic.gender.unique()

array(['F', 'Male', 'Female', 'U', 'Femal', 'M'], dtype=object)

#### Lets correct the column

In [22]:
CustomerDemographic['gender'] = CustomerDemographic['gender'].map(lambda x : 'Female' if x == 'F' or x == "Femal" else ('Male' if x == 'M' else ( "Unspecified" if x == 'U' else x)))

#### checking the gender column


In [23]:
CustomerDemographic["gender"].value_counts()

Female         2039
Male           1873
Unspecified      88
Name: gender, dtype: int64

#### Lets Check the null value

In [24]:
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

#### Lets remove Default from data set 

In [25]:
CustomerDemographic.drop('default',axis = 1,inplace = True)

#### Examining the data set

In [26]:
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

#### Examining  the deceased_indicator column

In [27]:
CustomerDemographic.deceased_indicator.value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

In [28]:
# Create a data frame for deceased customer

In [29]:
Deceased = CustomerDemographic[CustomerDemographic['deceased_indicator']== 'Y']

In [30]:
Deceased.head()

customer_id first_name    last_name  gender  \
752           753       Josy  St. Quentin  Female   
3789         3790     Kurtis       Morson    Male   

      past_3_years_bike_related_purchases        DOB      job_title  \
752                                    82 1970-03-07   Food Chemist   
3789                                   91 1959-05-31  Senior Editor   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
752                 Health  Affluent Customer                  Y      Yes   
3789                Retail      Mass Customer                  Y      Yes   

      tenure  
752      6.0  
3789     9.0

##### droping the Deceased index


In [31]:
CustomerDemographic.drop(Deceased.index,axis = 0,inplace = True)

#### checking the value count of deceased indicator 

In [32]:
CustomerDemographic['deceased_indicator'].value_counts()

N    3998
Name: deceased_indicator, dtype: int64

#### Examining the job_industry_category column


In [33]:
CustomerDemographic.job_industry_category.value_counts()

Manufacturing         799
Financial Services    774
Health                601
Retail                357
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

#### Correct the Agriculture spelling Mistake

In [34]:
CustomerDemographic['job_industry_category'] = CustomerDemographic['job_industry_category'].replace('Argiculture','Agriculture')

#### Lets check the job_industry_category Column

In [35]:
CustomerDemographic.job_industry_category.value_counts()

Manufacturing         799
Financial Services    774
Health                601
Retail                357
Property              267
IT                    223
Entertainment         136
Agriculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [36]:
CustomerDemographic.job_industry_category.value_counts(normalize = True)*100

Manufacturing         23.907840
Financial Services    23.159785
Health                17.983244
Retail                10.682226
Property               7.989228
IT                     6.672651
Entertainment          4.069420
Agriculture            3.381209
Telecommunications     2.154399
Name: job_industry_category, dtype: float64

#### Here we need to foucus on top for data values
1. Manufacturing         23.907840
2. Financial Services    23.159785
3. Health                17.983244
4. Retail                10.682226

#### So we can fill  the data with these values  

#### Lets check the percentage of null value in job_industry_category

In [37]:
CustomerDemographic.job_industry_category.isnull().mean()*100

16.408204102051023

#### As per value count and percentage of the missing value we can use following method :
1. As the percentage for  Manufacturing  is 23.907840 and is Financial Services  23.159785 
2. So we can fill 246 rows with Manufacturing and other 246 with Financial Services   
3. Remaining 164 we can split the half with Health and Retail

#### Lets create a new data frame for null rows of job_industry_category 

In [38]:
Customer_null =  CustomerDemographic[CustomerDemographic.job_industry_category.isnull()]

#### Lets check the data frame Customer_null

In [39]:
Customer_null.isnull().mean()*100

customer_id                              0.000000
first_name                               0.000000
last_name                                3.810976
gender                                   0.000000
past_3_years_bike_related_purchases      0.000000
DOB                                      0.000000
job_title                               16.006098
job_industry_category                  100.000000
wealth_segment                           0.000000
deceased_indicator                       0.000000
owns_car                                 0.000000
tenure                                   0.000000
dtype: float64

#### Lets  fill the data with our approach

##### Create a list of values to fill



In [40]:
fill_values = ['Manufacturing'] * 246 + ['Financial Services'] * 246 + ['Health'] * 82 + ['Retail'] * 82

##### Fill the null values




In [41]:
Customer_null['job_industry_category'].fillna(value=pd.Series(fill_values, index=Customer_null.index), inplace=True)

##### Verify the count of null values is zero


In [42]:
Customer_null.isnull().sum()

customer_id                              0
first_name                               0
last_name                               25
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                      0
job_title                              105
job_industry_category                    0
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
dtype: int64

#### Lets Drop Customer_null from main data frame


In [43]:
CustomerDemographic.drop(Customer_null.index,inplace = True)

#### Lets Check the shape

In [44]:
CustomerDemographic.shape

(3342, 12)

#### Lets Append the Customer_null in the data frame 

In [45]:
CustomerDemographic = CustomerDemographic.append(Customer_null)

#### Lets Recheck the data frame

In [46]:
CustomerDemographic.shape

(3998, 12)

#### Lets Fill Tenure Column 

##### filling the tenure data set with  median


In [47]:
tenure_median = CustomerDemographic['tenure'].median()

In [48]:
CustomerDemographic['tenure'].fillna(tenure_median,inplace = True)

#### Examine the data set

In [49]:
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                    0
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
dtype: int64

####  Dropping  the null job_title rows

In [50]:
CustomerDemographic['job_title']

0          Executive Secretary
1       Administrative Officer
2           Recruiting Manager
3                          NaN
5                          NaN
                 ...          
3965               Geologist I
3973          Technical Writer
3975          Internal Auditor
3982          Graphic Designer
3999      Software Engineer IV
Name: job_title, Length: 3998, dtype: object

#### -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

#### Lets check the CustomerAddress data frame

In [51]:
CustomerAddress.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

#### Examine the data set

In [52]:

CustomerAddress.shape

(3999, 6)

In [53]:
CustomerAddress.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [54]:
CustomerAddress.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

#### Lets check null values

In [55]:
CustomerAddress.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

#### Lets Check unique values

In [56]:
CustomerAddress['address'].unique()

array(['060 Morning Avenue', '6 Meadow Vale Court', '0 Holy Cross Court',
       ..., '87 Crescent Oaks Alley', '8194 Lien Street',
       '320 Acker Drive'], dtype=object)

In [57]:
CustomerAddress['state'].unique()

array(['New South Wales', 'QLD', 'VIC', 'NSW', 'Victoria'], dtype=object)

#### Lets replace the values

In [58]:
CustomerAddress['state'].replace('QLD','Queensland',inplace = True)
CustomerAddress['state'].replace('VIC','Victoria',inplace = True)
CustomerAddress['state'].replace('NSW','New South Wales',inplace = True)

#### Lets check the data set

In [59]:
CustomerAddress['state'].unique()

array(['New South Wales', 'Queensland', 'Victoria'], dtype=object)

In [60]:
CustomerAddress['country'].unique()

array(['Australia'], dtype=object)

#### Exporting dataframe to csv file in local system


In [67]:
# create an ExcelWriter object to write to a single Excel file
with pd.ExcelWriter('C:\\Users\\sanje\\OneDrive\\Desktop\\merged_data.xlsx') as writer:
    
    # write the transaction data to a sheet named 'Transaction'
    transaction.to_excel(writer, sheet_name='Transaction', index=False)
    
    # write the CustomerDemographic data to a sheet named 'CustomerDemographic'
    CustomerDemographic.to_excel(writer, sheet_name='CustomerDemographic', index=False)
    
    # write the CustomerAddress data to a sheet named 'CustomerAddress'
    CustomerAddress.to_excel(writer, sheet_name='CustomerAddress', index=False)